digiwf-engine/digiwf-engine-service/src/main/java/de/muenchen/oss/digiwf/cleanup/services/CleanupProcessDefinitionService.java

```diff
@@ -0,0 +1,111 @@
+                } catch (Exception e) {
+                    log.error("Failed to automatically migrate from process definition {} from {}", key, sourceDefinition.version());
+                }
+            }
+        );
+    }
+
+    public void deleteObviousDefinitions(String key, boolean ignoreHistorical) {
+        var definitions = serviceDefinitionService.getProcessDefinitionsWithInstanceInfoByKey(key);
+        var thresholdDate = Instant.now().minus(180, ChronoUnit.DAYS);
+        var forDeletion = definitions
+            .stream()
+            .filter(definitionWithInstanceInfo -> isObviousForDeletion(definitionWithInstanceInfo, thresholdDate, ignoreHistorical))
+            .map(ProcessDefinitionWithInstanceInfo::processDefinitionId)
+            .toList();
+        var remaining = definitions.stream().map(ProcessDefinitionWithInstanceInfo::processDefinitionId).filter(
+            definitionWithInstanceInfo -> !forDeletion.contains(definitionWithInstanceInfo)
+        ).toList();
+
+        deleteProcessDefinitions(key, forDeletion, remaining);
+    }
+
+
+    public void deleteAboveThreshold(String key, Integer thresholdCount) {
+        var definitions = serviceDefinitionService.getProcessDefinitionsWithInstanceInfoByKey(key);
+        if (definitions.size() <= thresholdCount) {
+            // nothing to do
+            return;
+        }
+        var forDeletion = definitions
+            .stream()
+            .limit(definitions.size() - thresholdCount) // take the definitions with lower version in order to get the last <thresholdCount> remaining
+            .map(ProcessDefinitionWithInstanceInfo::processDefinitionId)
+            .toList();
+        var remaining = definitions.stream().map(ProcessDefinitionWithInstanceInfo::processDefinitionId).filter(
+            def -> !forDeletion.contains(def)
+        ).toList();
+
+        deleteProcessDefinitions(key, forDeletion, remaining);
+    }
+
+    private void deleteProcessDefinitions(String key, List<String> forDeletion, List<String> remaining) {
+        log.info("Deleting definitions for key {}: \n[\n{}\n] \nand left over: \n[\n{}\n].",
+            key,
+            String.join(",\n", forDeletion),
+            String.join(",\n", remaining));
+        //
+        serviceDefinitionService.deleteDefinitions(true, forDeletion.toArray(new String[0]));
+    }
+
+
+    public boolean isObviousForDeletion(ProcessDefinitionWithInstanceInfo info, Instant thresholdDate, boolean ignoreHistorical) {
+        if (info.isLatest()) {
+            return false; // never delete latest definition
+        }
+        if (ignoreHistorical) {
+            return info.instanceCount() == 0; // delete old definitions without instances.
+        } else {
+            return info.newestProcessInstanceStartTime() != null && info.newestProcessInstanceStartTime().toInstant().isBefore(thresholdDate);
+        }
```

Comment:
Wenn `ignoreHistorical=true`, werden bis auf die letzte zum Löschen markiert, die keine Prozessinstanz haben.
Wenn `ignoreHistorical=false`, werden nur Prozessdefinitionen gelöscht, die alte Processinstanzen haben. Wenn in diesem Fall keine Prozessinstanz vorliegt, wird auch die Prozessdefinition nicht gelöscht. Sollte sie aber.

```suggestion
       if (info.instanceCount() == 0) return true;
       if (ignoreHistorical) return false;
       return info.newestProcessInstanceStartTime() != null && info.newestProcessInstanceStartTime().toInstant().isBefore(thresholdDate);
```